### PR TITLE
Benchmark Recipe

### PR DESCRIPTION
**Pull request by heitorschueroff**
This PR adds a recipe for PyTorch benchmark module.

TODO

- [x] Fix output formatting

cc @robieta, @brianjo, @ngimel

### CODE REVIEW

**Comment by heitorschueroff** [comment]
@brianjo Is there a better way to display output from code besides putting it in a code block? Something like what you'd get from a python notebook?

File: recipes_source/recipes/benchmark.py [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+######################################################################
+# 6. Generating bechmark inputs with Fuzzed Parameters
+#
+# We'll take a brief detour and use fuzzed inputs to discuss transpose
+# and contiguous before returning to shift.
+#
+
+from torch.utils.benchmark import Fuzzer, FuzzedParameter, FuzzedTensor, ParameterAlias
+
+example_fuzzer = Fuzzer(
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:397]: a description of what this fuzzer does (generates 2d tensors with sizes distributed according to specified distributions) is useful. You are talking to people who see Fuzzer and FuzzedParameter for the first time.

File: recipes_source/recipes/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+# Times are in nanoseconds (ns).
+
+######################################################################
+# The fuzzed benchmarks reveal several noteworthy features:
+#
+# If a Tensor is already contiguous we do not need to construct a new
+# Tensor and the operation is extremely cheap. O(100 ns)
+#
+# For N x 1 Tensors transpose requires that we create a new Tensor,
+# but we can reuse the same buffer.
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:448]: `buffer` is an unclear and overloaded word, "we don't have to do any physical data movement"

File: recipes_source/recipes/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,828 @@
...
+#
+# For N x N tensors, either contiguous or transposed contiguous will
+# be expensive depending on the underlying data layout.
+
+######################################################################
+# 6.1. Canned fuzzers: back to our x+1 kernel
+#
+# When benchmarking an op, there are a lot of things to consider:
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:456]: yeah, something along these lines should be in the beginning of section 6, and it's ok to repeat it here, now introducing fuzzers for unary and binary ops.

File: recipes_source/recipes/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,828 @@
...
+#
+
+from torch.utils.benchmark.op_fuzzers import unary
+
+def skip_if(x):
+    # We know that around this point our custom kernel is slower than vanilla add, so
+    # there's not much useful signal from testing larger Tensors.
+    return x.numel() > 64 * 1024
+
+def to_description(x, x_params):
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:470]: comment what this function does

File: recipes_source/recipes/benchmark.py [hunk 5]
```diff
@@ -0,0 +1,828 @@
...
+# [5]    [325, 128]                                        [::2, :]
+# [6]    [118, 166]                                        
+# [7]    [6309]                                            
+# [8]    [26, 512]                                         
+# [9]    [27, 225]           (1, 0)                        
+# [10]   [380, 58]                                         
+
+######################################################################
+# There's an interesting pattern in our benchmark results above. The
+# dicontiguous input tensors [1], [2], [5] and [9] take much longer
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:543]: "discontiguous"

File: recipes_source/recipes/benchmark.py [hunk 6]
```diff
@@ -0,0 +1,828 @@
...
+# to run on than expected. The reason is that we are cloning the
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:544]: you don't say which implementation takes much longer to run, and also how do you know what to expect?

**Review by ngimel — COMMENTED** [review]
This tutorial feels like 2 in 1 - basic use of benchmarking tools, and how to optimize your custom op implementation. Many people to whom benchmarking tools are addressed don't write a single line of c++ in their daily lives, and yet would still benefit from knowing how to benchmark their code correctly. So it might make sense to either split it in 2, or structure it in such a way so that there's a first part for everyone, and second part for those who are still able to follow. 
I didn't get to the end, but let me know if it makes sense.

File: recipes_source/recipes/README.txt [hunk 7]
```diff
@@ -55,8 +55,16 @@ PyTorch Recipes
 
 14. mobile_perf.py
          PyTorch Mobile Performance Recipes
-         https://pytorch.org/tutorials/recipes/mobile_perf.html
+         https://pytorch.org/tutorials/recipes/recipes/mobile_perf.html
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply recipes_source/recipes/README.txt:58]: The actual link has recipes twice in it.

[the remark this answers is not in the document]
Comment (heitorschueroff) [reply recipes_source/recipes/README.txt:58]: When I was testing the links had recipes twice in it, but now I'm trying it and it was right before. I'll fix it.

File: recipes_source/recipes/benchmark.py [hunk 8]
```diff
@@ -0,0 +1,828 @@
...
+# 2. Writing the custom C++ extension
+# 3. Simple benchmark using timeit.Timer
+# 4. Better benchmark using torch.benchmark.Timer
+# 4.1. Timer.blocked_autorage
+# 4.2. Why runtime awareness matters
+# 5. Comparing different benchmarks
+# 6. Generating bechmark inputs with Fuzzed Parameters
+# 6.1. Canned fuzzers: back to our x+1 kernel
+# 7. Taking more precise measurements with instruction counting
+#
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply recipes_source/recipes/benchmark.py:55]: I agree with you! We will update it so the first few steps are targeted to general users and we will keep it python only. Later when looking into callgrind stats we might add a custom C++ op.

File: recipes_source/recipes/benchmark.py [hunk 9]
```diff
@@ -0,0 +1,765 @@
...
+    cpp_sources=batched_dot_src,
+    extra_cflags=['-O3'],
+    extra_include_paths=[
+        # `load_inline` needs to know where to find Pybind11 headers.
+        os.path.join(os.getenv('CONDA_PREFIX'), 'include')
+    ],
+    functions=['batched_dot_mul_sum_v0', 'batched_dot_mul_sum_v1']
+)
+
+module_import_str = f"""\
```
Comment (ngimel) [inline recipes_source/recipes/benchmark.py:612]: a comment on what's going on here?

**Review by ngimel — APPROVED** [review]
python portion looks good now! C++ section could use some updating once latest improvements that avoid python altogether are landed.

**Pull request merged**